orbit_predictor/predictors/base.py

```diff
@@ -279,6 +281,76 @@ def get_next_pass(self, location, when_utc=None, max_elevation_gt=5,
         else:
             raise NotReachable('Propagation limit date exceeded')
 
+    def eclipses_since(self, when_utc, limit_date=None):
+        """
+        An iterator that yields all eclipses start and end times between
+        when_utc and limit_date.
+
+        The next eclipse with a end strictly after when_utc will be returned,
+        possibly the current eclipse.
+        The last eclipse returned starts before limit_date, but it can end
+        strictly after limit_date.
+        No circular orbits are not supported, and will raise NotImplementedError.
+        """
+        def _get_illumination(t):
+            my_start = start + dt.timedelta(seconds=t)
+            result = get_satellite_minus_penumbra_verticals(
+                self.get_only_position(my_start),
+                my_start
+            )
+            return result
+
+        orbital_period_s = 60 / self.mean_motion * 2 * np.pi
+        # A third orbit is used as the base window of the search.
+        # This window ensures the function get_satellite_minus_penumbra_verticals
+        # will not have more than one local minimum (one in the illuminated phase and
+        # the other in penumbra).
+        base_search_window_s = orbital_period_s / 3
+        start = when_utc
+
+        while limit_date is None or start < limit_date:
+
+            # a minimum negative value is aproximatelly the middle point of the eclipse
+            minimum_illumination = minimize_scalar(
+                _get_illumination,
+                bounds=(0, base_search_window_s),
+                method="bounded",
+                options={"xatol": 1e-2},
+            )
+            eclipse_center_candidate_delta_s = minimum_illumination.x
+
+            # If found a minimum that is not illuminated, there is an eclipse here
+            if _get_illumination(eclipse_center_candidate_delta_s) < 0:
+                # The small time interval to search zeros around the center
+                # is estimated with the expected eclipse duration (which generally
+                # is smaller than expected, and that is the reason of the 1.5 coeficient).
+                # Also a minimum of 180 seconds was added because
+                # in some cases the estimation is 0 even though there is an eclipse.
+                eclipse_duration_estimation_s = self.get_eclipse_duration(start) * 60
+                zero_search_window_s = max(180, 1.5 * eclipse_duration_estimation_s)
+
+                # Search now both zeros to get the start and end of the eclipse
+                eclipse_start_delta_s = brentq(
+                    _get_illumination,
+                    eclipse_center_candidate_delta_s - zero_search_window_s,
+                    eclipse_center_candidate_delta_s,
+                    xtol=1e-2,
+                    full_output=False,
+                )
+                eclipse_end_delta_s = brentq(
```

Comment:
Can we assume that `eclipse_end_delta_s = eclipse_start_delta_s + dt.timedelta(minutes=self.get_eclipse_duration(eclipse_start_delta_s))` ? I know it's an approximation, but we save iterating over the function again.

Another idea is to use `eclipse_start_delta_s + dt.timedelta(minutes=self.get_eclipse_duration(eclipse_start_delta_s))` as a starting point of a non-bracketed method, like `scipy.optimize.newton`.

Reply:
The first assumption cannot be made in all cases. As far as I checked, the `get_eclipse_duration` returns durations that can differ in a few minutes from the one obtained with this method. In the 6 AM LTAN orbit, it returned 0 eclipse duration where there was still some eclipses. But perhaps we should open an issue to solve this difference instead of avoiding using the function.

The other Idea it's great. I'll check if the performance is improved. If there is not much gain, I think we should keep it as it is now, since it is symmetric with the search of the start, and so, the code is easier to understand.

Reply:
> But perhaps we should open an issue to solve this difference instead of avoiding using the function.

https://github.com/satellogic/orbit-predictor/issues/92

Not necessarily a blocker for this PR if it takes too much time to fix, but I prefer to note it.